src/mds/Capability.cc

```diff
@@ -183,7 +183,7 @@ client_t Capability::get_client() const
 
 int Capability::confirm_receipt(ceph_seq_t seq, unsigned caps) {
   int was_revoking = (_issued & ~_pending);
-  if (seq == last_sent) {
+  if (seq <= last_sent) {
```

Comment:
This cold is relatively ancient in the MDS:

https://github.com/ceph/ceph/commit/35bac67a605a11bbbadf4c247c37468ec5dfbc6f#diff-e108a3d821c93ed8411c681edfe5adff57ef41ef021e83e0a769aa6c8e6bbcc9R167

which makes me **very** concerned about changing it. I tried looking more into the MDS logs:

```
2023-06-15T06:33:23.293+0000 7fde93b16700  1 -- [v2:172.21.15.29:6832/1469037096,v1:172.21.15.29:6833/1469037096] --> 192.168.0.1:0/1731885576 -- client_caps(revoke ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcb dirty=- wanted=Fcr follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 -- 0x556a82154380 con 0x556a81cc5c00
2023-06-15T06:33:23.297+0000 7fde93b16700  1 -- [v2:172.21.15.29:6832/1469037096,v1:172.21.15.29:6833/1469037096] <== client.4941 192.168.0.1:0/1731885576 4979 ==== client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcrb dirty=- wanted=Fcr follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 ==== 260+0+0 (crc 0 0 0) 0x556a82154380 con 0x556a81cc5c00
2023-06-15T06:33:23.297+0000 7fde93b16700  1 -- [v2:172.21.15.29:6832/1469037096,v1:172.21.15.29:6833/1469037096] <== client.4941 192.168.0.1:0/1731885576 4980 ==== client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFc dirty=- wanted=- follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 ==== 260+0+0 (crc 0 0 0) 0x556a8282d180 con 0x556a81cc5c00
```

From: /teuthology/vshankar-2023-06-15_04:58:28-fs-wip-vshankar-testing-20230614.124123-testing-default-smithi/7305053/remote/smithi029/log/ceph-mds.a.log.gz


For the client:

```
2023-06-15T06:33:23.293+0000 7fbdd2ffd700  1 -- 192.168.0.1:0/1731885576 <== mds.2 v2:172.21.15.29:6832/1469037096 10842 ==== client_caps(revoke ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcb dirty=- wanted=Fcr follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 ==== 260+0+0 (crc 0 0 0) 0x7fbde403b040 con 0x7fbde402b2b0
2023-06-15T06:33:23.293+0000 7fbdd2ffd700 10 client.4941  mds.2 seq now 6709
2023-06-15T06:33:23.293+0000 7fbdd2ffd700  5 client.4941 handle_cap_grant on in 0x3000000025b mds.2 seq 18 caps now pAsLsXsFcb was pAsLsXsFxcrb
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941 update_inode_file_time 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=1,2048=1,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFxcrb(2=pAsLsXsFxcrb) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) pAsLsXsFxcrb ctime 2023-06-15T06:33:20.586624+0000 mtime 2023-06-15T06:33:20.586624+0000                                                                                                                                                       
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941   revocation of Fxr
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941 check_caps on 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=1,2048=1,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFcb(2=pAsLsXsFcb) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) wanted Fcr used Fcr issued pAsLsXsFcb revoking Fxr flags=1
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941  cap mds.2 issued pAsLsXsFcb implemented pAsLsXsFxcrb revoking Fxr
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941 send_cap 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=1,2048=1,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFcb(2=pAsLsXsFcb) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) mds.2 seq 18 used Fcr want Fcr flush - retain pAsxLsxXsxFscwbl held pAsLsXsFxcrb revoking Fxr dropping -                                                                                                                                                         
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 15 client.4941 auth cap, reset requested_max_size due to not wanting any file write cap
2023-06-15T06:33:23.297+0000 7fbde3fff700  1 -- 192.168.0.1:0/1731885576 <== osd.4 v2:172.21.15.47:6808/4047423328 1723 ==== osd_op_reply(13867 3000000025b.00000001 [read 0~1790296 out=1790296b] v0'0 uv340 ondisk = 0) v8 ==== 164+0+1790296 (crc 0 0 0) 0x7fbdd58c0440 con 0x7fbdb423fc70                                                                                                 
2023-06-15T06:33:23.297+0000 7fbdd2ffd700  1 -- 192.168.0.1:0/1731885576 --> [v2:172.21.15.29:6832/1469037096,v1:172.21.15.29:6833/1469037096] -- client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFcrb dirty=- wanted=Fcr follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 -- 0x7fbd690bbe00 con 0x7fbde402b2b0
2023-06-15T06:33:23.297+0000 7fbd73fff700  3 client.4941 ll_lookup 0x1000000314d.head f83
2023-06-15T06:33:23.297+0000 7fbd937fe700  3 client.4941 ll_read 0x7fbda85ac3f0 0x300000002ee  2621440~131072
2023-06-15T06:33:23.297+0000 7fbdd3fff700 20 client.4941 C_Readahead on 0x7fbdcd494930
2023-06-15T06:33:23.297+0000 7fbdd3fff700  5 client.4941 put_cap_ref dropped last FILE_CACHE ref on 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFcb(2=pAsLsXsFcb) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930)
2023-06-15T06:33:23.297+0000 7fbdd3fff700 10 client.4941 check_caps on 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFcb(2=pAsLsXsFcb) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) wanted - used Fc issued pAsLsXsFcb revoking Fr flags=0
2023-06-15T06:33:23.297+0000 7fbdd3fff700 10 client.4941  cap mds.2 issued pAsLsXsFcb implemented pAsLsXsFcrb revoking Fr
2023-06-15T06:33:23.297+0000 7fbdd3fff700 10 client.4941 completed revocation of Fr
2023-06-15T06:33:23.297+0000 7fbdd3fff700 10 client.4941 send_cap 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFcb(2=pAsLsXsFcb) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) mds.2 seq 18 used Fc want - flush - retain pAsLsXsFsc held pAsLsXsFcrb revoking Fr dropping Fb                                                                                                                                                                   
2023-06-15T06:33:23.297+0000 7fbdd3fff700 15 client.4941 auth cap, reset requested_max_size due to not wanting any file write cap
2023-06-15T06:33:23.297+0000 7fbdd3fff700  1 -- 192.168.0.1:0/1731885576 --> [v2:172.21.15.29:6832/1469037096,v1:172.21.15.29:6833/1469037096] -- client_caps(update ino 0x3000000025b 1 seq 18 caps=pAsLsXsFc dirty=- wanted=- follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 -- 0x7fbda4002770 con 0x7fbde402b2b0
2023-06-15T06:33:23.297+0000 7fbdd3fff700 20 client.4941 put_inode on 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFc(2=pAsLsXsFc) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) n = 1
2023-06-15T06:33:23.297+0000 7fbdd3fff700 20 client.4941 put_inode on 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFc(2=pAsLsXsFc) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) n = 1
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _ll_get 0x7fbdcd3a9690 0x100000030a8 -> 3
2023-06-15T06:33:23.297+0000 7fbd91ffb700  3 client.4941 ll_lookup 0x100000030a8.head da
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 may_lookup 0x100000030a8.head(faked_ino=0 nref=4 ll_ref=3 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:56.818299+0000 mtime=2023-06-15T06:33:00.118373+0000 ctime=2023-06-15T06:33:00.118373+0000 change_attr=16 caps=pAsLsXsFsx(0=pAsLsXsFsx) COMPLETE parents=0x100000030a4.head["p1"] 0x7fbdcd3a9690); UserPerm(uid: 1000, gid: 1267)
2023-06-15T06:33:23.297+0000 7fbd91ffb700 10 client.4941 _getattr mask As issued=1
2023-06-15T06:33:23.297+0000 7fbd91ffb700  3 client.4941 may_lookup 0x7fbdcd3a9690 = 0
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _lookup have 0x100000030a8.head["da"] from mds.-1 ttl 0.000000 seq 0
2023-06-15T06:33:23.297+0000 7fbd91ffb700 10 client.4941 _lookup 0x100000030a8.head(faked_ino=0 nref=4 ll_ref=3 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:56.818299+0000 mtime=2023-06-15T06:33:00.118373+0000 ctime=2023-06-15T06:33:00.118373+0000 change_attr=16 caps=pAsLsXsFsx(0=pAsLsXsFsx) COMPLETE parents=0x100000030a4.head["p1"] 0x7fbdcd3a9690) da = 0x10000003177.head(faked_ino=0 nref=5 ll_ref=779 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:59.711524+0000 mtime=2023-06-15T06:33:21.524249+0000 ctime=2023-06-15T06:33:21.524249+0000 change_attr=27 caps=pAsLsXsFsx(0=pAsLsXsFsx) parents=0x100000030a8.head["da"] 0x7fbdcd40d8c0)                                                                               
2023-06-15T06:33:23.297+0000 7fbd91ffb700 10 client.4941 fill_stat on 0x10000003177 snap/devhead mode 040777 mtime 2023-06-15T06:33:21.524249+0000 ctime 2023-06-15T06:33:21.524249+0000
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _ll_get 0x7fbdcd40d8c0 0x10000003177 -> 780
2023-06-15T06:33:23.297+0000 7fbd91ffb700  3 client.4941 ll_lookup 0x100000030a8.head da -> 0 (10000003177)
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 put_inode on 0x10000003177.head(faked_ino=0 nref=5 ll_ref=780 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:59.711524+0000 mtime=2023-06-15T06:33:21.524249+0000 ctime=2023-06-15T06:33:21.524249+0000 change_attr=27 caps=pAsLsXsFsx(0=pAsLsXsFsx) parents=0x100000030a8.head["da"] 0x7fbdcd40d8c0) n = 1  
2023-06-15T06:33:23.297+0000 7fbd91ffb700  8 client.4941 _ll_forget 0x100000030a8 1
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _ll_put 0x7fbdcd3a9690 0x100000030a8 1 -> 2
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _ll_get 0x7fbdcd3a8990 0x100000030a6 -> 3
2023-06-15T06:33:23.297+0000 7fbd727fc700 20 client.4941 _ll_get 0x7fbdcd40d8c0 0x10000003177 -> 781
2023-06-15T06:33:23.297+0000 7fbd91ffb700  3 client.4941 ll_lookup 0x100000030a6.head d5
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 may_lookup 0x100000030a6.head(faked_ino=0 nref=4 ll_ref=3 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:56.811910+0000 mtime=2023-06-15T06:32:58.982029+0000 ctime=2023-06-15T06:32:58.982029+0000 change_attr=9 caps=pAsLsXs(0=pAsLsXs,2=pAsLsXs) parents=0x100000030a4.head["p0"] 0x7fbdcd3a8990); UserPerm(uid: 1000, gid: 1267)
2023-06-15T06:33:23.297+0000 7fbd91ffb700 10 client.4941 _getattr mask As issued=1
2023-06-15T06:33:23.297+0000 7fbd91ffb700  3 client.4941 may_lookup 0x7fbdcd3a8990 = 0
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _lookup have 0x100000030a6.head["d5"] from mds.2 ttl 2023-06-15T06:33:51.553844+0000 seq 20
2023-06-15T06:33:23.297+0000 7fbd727fc700  3 client.4941 ll_lookup 0x10000003177.head db
2023-06-15T06:33:23.297+0000 7fbd91ffb700 10 client.4941 _lookup 0x100000030a6.head(faked_ino=0 nref=4 ll_ref=3 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:56.811910+0000 mtime=2023-06-15T06:32:58.982029+0000 ctime=2023-06-15T06:32:58.982029+0000 change_attr=9 caps=pAsLsXs(0=pAsLsXs,2=pAsLsXs) parents=0x100000030a4.head["p0"] 0x7fbdcd3a8990) d5 = 0x100000030cd.head(faked_ino=0 nref=5 ll_ref=708 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:56.906352+0000 mtime=2023-06-15T06:33:15.497787+0000 ctime=2023-06-15T06:33:15.497787+0000 change_attr=33 caps=pAsLsXsFsx(2=pAsLsXsFsx) parents=0x100000030a6.head["d5"] 0x7fbdcd3b9470)                                                                                     
2023-06-15T06:33:23.297+0000 7fbd91ffb700 10 client.4941 fill_stat on 0x100000030cd snap/devhead mode 040777 mtime 2023-06-15T06:33:15.497787+0000 ctime 2023-06-15T06:33:15.497787+0000
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 _ll_get 0x7fbdcd3b9470 0x100000030cd -> 709
2023-06-15T06:33:23.297+0000 7fbd91ffb700  3 client.4941 ll_lookup 0x100000030a6.head d5 -> 0 (100000030cd)
2023-06-15T06:33:23.297+0000 7fbd91ffb700 20 client.4941 put_inode on 0x100000030cd.head(faked_ino=0 nref=5 ll_ref=709 cap_refs={} open={} mode=40777 size=0/0 nlink=1 btime=2023-06-15T06:32:56.906352+0000 mtime=2023-06-15T06:33:15.497787+0000 ctime=2023-06-15T06:33:15.497787+0000 change_attr=33 caps=pAsLsXsFsx(2=pAsLsXsFsx) parents=0x100000030a6.head["d5"] 0x7fbdcd3b9470) n = 1  
2023-06-15T06:33:23.297+0000 7fbd937fe700 10 client.4941 get_caps 0x300000002ee.head(faked_ino=0 nref=13 ll_ref=7 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=1,2=0} mode=100666 size=4194304/12582912 nlink=1 btime=2023-06-15T06:33:05.778358+0000 mtime=2023-06-15T06:33:20.378024+0000 ctime=2023-06-15T06:33:20.378024+0000 change_attr=36 caps=pAsLsXsFsxcrwb(2=pAsLsXsFsxcrwb) objectset[0x300000002ee ts 0/0 objects 1 dirty_or_tx 0] parents=0x10000003135.head["f4c"] 0x7fbdcd4eea10) have pAsLsXsFsxcrwb need Fr want Fc revoking -
2023-06-15T06:33:23.297+0000 7fbdd2ffd700  1 -- 192.168.0.1:0/1731885576 <== mds.2 v2:172.21.15.29:6832/1469037096 10843 ==== client_caps(revoke ino 0x3000000025b 1 seq 19 caps=pAsLsXsFc dirty=- wanted=Fcr follows 0 size 5984600/12582912 ts 1/18446744073709551615 mtime 2023-06-15T06:33:20.586624+0000 ctime 2023-06-15T06:33:20.586624+0000 change_attr 136 tws 1) v12 ==== 260+0+0 (crc 0 0 0) 0x7fbda4002770 con 0x7fbde402b2b0
...
2023-06-15T06:33:23.297+0000 7fbdd2ffd700  5 client.4941 handle_cap_grant on in 0x3000000025b mds.2 seq 19 caps now pAsLsXsFc was pAsLsXsFc
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941 update_inode_file_time 0x3000000025b.head(faked_ino=0 nref=20 ll_ref=23 cap_refs={4=0,1024=0,2048=0,4096=0,8192=0} open={1=0,2=0} mode=100666 size=5984600/12582912 nlink=2 btime=2023-06-15T06:33:01.855043+0000 mtime=2023-06-15T06:33:20.586624+0000 ctime=2023-06-15T06:33:20.586624+0000 change_attr=136 caps=pAsLsXsFc(2=pAsLsXsFc) objectset[0x3000000025b ts 0/0 objects 2 dirty_or_tx 0] parents=0x3000000024f.head["f28"], 0x30000000255.head["f2c"] 0x7fbdcd494930) pAsLsXsFc ctime 2023-06-15T06:33:20.586624+0000 mtime 2023-06-15T06:33:20.586624+0000
2023-06-15T06:33:23.297+0000 7fbdd2ffd700 10 client.4941   caps unchanged at pAsLsXsFc
```

From: /teuthology/vshankar-2023-06-15_04:58:28-fs-wip-vshankar-testing-20230614.124123-testing-default-smithi/7305053/remote/smithi029/log/ceph-client.0.41548.log.gz

So the interesting bit from the client log is that it was in the middle of an async read which blocked the release of `Frc` caps. That happened later once the cap ref was dropped. Consequently, there were two cap updates for that revoke.

The issue, AFAICT, is here:

https://github.com/ceph/ceph/blob/fd62a8dfcb345fabed3eea6b3a461f1ca848a239/src/client/Client.cc#L5811

This code ignores the actual cap op as part of `MClientCaps` and just decides whether it's a revoke based on a difference in new_caps/issued. I would contend that the Client should _always_ send an ack if the MDS revokes cap even if a prior cap update dropped the relevant cap.

Reply:
Yeah, this sounds more reasonable. I will update it. Thanks @batrick

Reply:
> Yeah, this sounds more reasonable. I will update it. Thanks @batrick

For discussion sake:

The 2 (two) MClientCaps update message with seq=18 sent by the client to mds(.2) were correctly handled by the MDS. The revoke message sent by the MDS with seq=19 wasn't acknowledged by the client, so, therefore the initial proposed fix

```
if (seq <= last_sent) {
    ...
}
```

wouldn't have fixed anything, right?

Reply:
Yeah, it will fix it. Though the client will do nothing with the revoke msg, with the above fix the mds will update the caps anyway.